File: vendkit/core/declaration.py
"""Export declaration: the single source of slice identity (DR-0002).

Schema: export-declaration spec v1. YAML is imported lazily (via util);
gate-path code never imports this module.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from pathlib import Path

from .util import UsageError, load_yaml, match_any

SLUG = re.compile(r"^[a-z][a-z0-9-]{0,15}$")
DEFAULT_DECL = "vendkit-export.yml"
ADAPTER_KINDS = ("prefix-namespace", "glob-localise")


@dataclass
class Adapter:
    kind: str
    match: str
    prefix: str = ""
    fm_field: str = ""
    catalogue: dict[str, list[str]] = field(default_factory=dict)


@dataclass
class Profile:
    name: str
    export_include: list[str] = field(default_factory=lambda: ["*"])
    export_exclude: list[str] = field(default_factory=list)


@dataclass
class ExportDecl:
    slice_name: str
    slice_title: str
    publisher_scm: str
    publisher_repo: str
    include: list[str]
    exclude: list[str]
    seed: list[str]
    adapters: list[Adapter]
    profiles: dict[str, Profile]
    retracted: list[str]
    manifest_name: str
    path: str = DEFAULT_DECL

    # -- loading -----------------------------------------------------------

    @classmethod
    def load(cls, path: str) -> "ExportDecl":
        data = load_yaml(path)
        errs: list[str] = []

        if data.get("schema_version") != 1:
            errs.append("schema_version must be 1")

        sl = data.get("slice") or {}
        name = sl.get("name", "")
        if not isinstance(name, str) or not SLUG.match(name):
            errs.append(f"slice.name {name!r} must match {SLUG.pattern}")
        title = sl.get("title") or name

        pub = data.get("publisher") or {}
        scm = pub.get("scm", "")
        if scm not in ("github", "azure-repos"):
            errs.append("publisher.scm must be 'github' or 'azure-repos'")
        repo = pub.get("repo", "")
        # A URL/path is used verbatim by git; shorthand is scm-expanded
        # (core/upstream.py). Core never branches on scm beyond that
        # expansion — it is provenance and link metadata (DR-0015).
        if not isinstance(repo, str) or not repo:
            errs.append("publisher.repo is required (git URL or shorthand)")

        include = data.get("include") or []
        seed = data.get("seed") or []
        if not all(isinstance(p, str) for p in include + seed):
            errs.append("include/seed must be lists of glob strings")
        if not include and not seed:
            errs.append("at least one of include/seed must be non-empty")
        exclude = data.get("exclude") or []

        adapters: list[Adapter] = []
        for i, raw in enumerate(data.get("adapters") or []):
            kind = raw.get("kind")
            if kind not in ADAPTER_KINDS:
                # Hard error, never a silent skip (DR-0009).
                errs.append(f"adapters[{i}]: unknown kind {kind!r}")
                continue
            adp = Adapter(kind=kind, match=raw.get("match", ""))
            if not adp.match:
                errs.append(f"adapters[{i}]: match is required")
            if kind == "prefix-namespace":
                adp.prefix = raw.get("prefix", "")
                if not adp.prefix:
                    errs.append(f"adapters[{i}]: prefix is required")
            else:
                adp.fm_field = raw.get("field", "")
                adp.catalogue = raw.get("catalogue") or {}
                if not adp.fm_field:
                    errs.append(f"adapters[{i}]: field is required")
            adapters.append(adp)

        profiles: dict[str, Profile] = {}
        for pname, praw in (data.get("profiles") or {}).items():
            praw = praw or {}
            es = praw.get("export_slice") or {}
            profiles[pname] = Profile(
                name=pname,
                export_include=es.get("include") or ["*"],
                export_exclude=es.get("exclude") or [],
            )

        retracted = data.get("retracted") or []
        manifest_name = data.get("manifest_name") or f"{name}-manifest.json"

        known = {"schema_version", "slice", "publisher", "include", "exclude",
                 "seed", "adapters", "profiles", "retracted", "manifest_name"}
        for key in data:
            if key not in known:
                errs.append(f"unknown top-level key: {key!r}")

        if errs:
            raise UsageError(f"{path}: " + "; ".join(errs))
        return cls(
            slice_name=name, slice_title=title,
            publisher_scm=scm, publisher_repo=repo,
            include=include, exclude=exclude, seed=seed, adapters=adapters,
            profiles=profiles, retracted=retracted,
            manifest_name=manifest_name, path=path,
        )

    # -- export surface ----------------------------------------------------

    def _matched(self, root: str, patterns: list[str]) -> list[str]:
        """Sorted repo-relative posix paths: matched(patterns) − matched(exclude).

        Regular files only; symlinks are rejected (export-declaration spec §2).
        """
        rootp = Path(root)
        found: set[str] = set()
        for pattern in patterns:
            for hit in rootp.glob(pattern):
                rel = hit.relative_to(rootp).as_posix()
                if hit.is_symlink():
                    raise UsageError(f"symlink in export surface: {rel}")
                if hit.is_file():
                    found.add(rel)
        return [p for p in sorted(found) if not match_any(p, self.exclude)]

    def exported_files(self, root: str) -> list[str]:
        """The vendored (drift-gated) surface."""
        result = self._matched(root, self.include)
        if not result and not self.seed:
            raise UsageError(f"{self.path}: export surface is empty")
        overlap = set(result) & set(self._matched(root, self.seed))
        if overlap:
            # A path cannot be both drift-gated and free-to-diverge (DR-0013).
            raise UsageError(
                f"{self.path}: path(s) matched by both include and seed: "
                f"{', '.join(sorted(overlap)[:3])}"
            )
        return result

    def seeded_files(self, root: str) -> list[str]:
        """The scaffold-once surface (DR-0013): materialised only when the
        consumer path does not exist, then consumer-owned."""
        return self._matched(root, self.seed)

    # -- adapters ----------------------------------------------------------

    def adapters_for(self, path: str) -> list[Adapter]:
        hits = [a for a in self.adapters if match_any(path, [a.match])]
        for kind in ADAPTER_KINDS:
            if sum(1 for a in hits if a.kind == kind) > 1:
                raise UsageError(
                    f"{path}: more than one {kind} adapter matches"
                )
        return hits

    def consumer_path(self, path: str) -> str:
        for a in self.adapters_for(path):
            if a.kind == "prefix-namespace":
                head, _, tail = path.rpartition("/")
                if not tail.startswith(a.prefix):
                    tail = a.prefix + tail
                return f"{head}/{tail}" if head else tail
        return path

    # -- profiles ----------------------------------------------------------

    def profile_in_scope(self, profile: str | None, path: str) -> bool:
        """Whether reconcile-scope may offer `path` to this profile."""
        if profile is None or profile not in self.profiles:
            return True  # unbound consumer takes the whole surface
        prof = self.profiles[profile]
        return match_any(path, prof.export_include) and not match_any(
            path, prof.export_exclude
        )

File: internal/core/declaration.go
// Export declaration: the single source of slice identity (DR-0002).
// Schema: export-declaration spec v1.

package core

import (
	"fmt"
	"path"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
)

var slugRx = regexp.MustCompile(`^[a-z][a-z0-9-]{0,15}$`)

// cleanManifestDir normalises publisher.manifest_dir to a clean repo-relative
// POSIX directory, or reports invalid. "" and "." both mean the repo root.
// (Package-level so the `path` import is not shadowed by LoadExportDecl's
// `path` parameter.)
func cleanManifestDir(raw string) (string, bool) {
	if raw == "" {
		return "", true
	}
	clean := path.Clean(strings.ReplaceAll(raw, `\`, "/"))
	if clean == "." {
		return "", true
	}
	if path.IsAbs(clean) || clean == ".." || strings.HasPrefix(clean, "../") {
		return "", false
	}
	return clean, true
}

// ManifestRepoRel: the publisher-side, repo-relative POSIX path of the generated
// manifest. The consumer-side path is always `.vendkit/<ManifestName>`, so
// ManifestDir is publisher-local and never travels with the slice.
func (d *ExportDecl) ManifestRepoRel() string {
	if d.ManifestDir == "" {
		return d.ManifestName
	}
	return path.Join(d.ManifestDir, d.ManifestName)
}

// PublisherManifestPath: filesystem path of the manifest under a publisher root.
func (d *ExportDecl) PublisherManifestPath(root string) string {
	return filepath.Join(root, filepath.FromSlash(d.ManifestRepoRel()))
}

const DefaultDecl = "vendkit-export.yml"

var AdapterKinds = []string{"prefix-namespace", "glob-localise"}

type Adapter struct {
	Kind      string
	Match     string
	Prefix    string
	FMField   string
	Catalogue map[string][]string
}

type Profile struct {
	Name          string
	ExportInclude []string
	ExportExclude []string
}

type ExportDecl struct {
	SliceName     string
	SliceTitle    string
	PublisherSCM  string
	PublisherRepo string
	// ManifestDir: publisher-local directory holding the generated manifest
	// (empty = repo root). Publisher-side only — see ManifestRepoRel.
	ManifestDir  string
	Include      []string
	Exclude      []string
	Seed         []string
	Adapters     []Adapter
	Profiles     map[string]Profile
	Retracted    []string
	ManifestName string
	Path         string
}

func LoadExportDecl(path string) (*ExportDecl, error) {
	data, err := LoadYAML(path)
	if err != nil {
		return nil, err
	}
	var errs []string

	if !schemaVersionIs(data, 1) {
		errs = append(errs, "schema_version must be 1")
	}
	sl := getMap(data, "slice")
	name := getStr(sl, "name")
	if !slugRx.MatchString(name) {
		errs = append(errs, fmt.Sprintf("slice.name %q must match %s", name, slugRx.String()))
	}
	title := getStr(sl, "title")
	if title == "" {
		title = name
	}

	pub := getMap(data, "publisher")
	scm := getStr(pub, "scm")
	if scm != "github" && scm != "azure-repos" {
		errs = append(errs, "publisher.scm must be 'github' or 'azure-repos'")
	}
	repo := getStr(pub, "repo")
	if repo == "" {
		errs = append(errs, "publisher.repo is required (git URL or shorthand)")
	}
	// Publisher-local manifest directory. Nested under `publisher:` deliberately:
	// the consumer-side location is NOT configurable — a vendored manifest always
	// lands in `.vendkit/<manifest_name>` (manifest-and-gate spec §2) — so this
	// key must not read as if it travelled with the slice.
	manifestDir, dirOK := cleanManifestDir(getStr(pub, "manifest_dir"))
	if !dirOK {
		errs = append(errs, fmt.Sprintf(
			"publisher.manifest_dir %q must be a relative path inside the repo",
			getStr(pub, "manifest_dir")))
	}

	include, ok1 := strList(data["include"])
	seed, ok2 := strList(data["seed"])
	if !ok1 || !ok2 {
		errs = append(errs, "include/seed must be lists of glob strings")
	}
	if len(include) == 0 && len(seed) == 0 {
		errs = append(errs, "at least one of include/seed must be non-empty")
	}
	exclude, _ := strList(data["exclude"])

	var adapters []Adapter
	for i, raw := range getList(data, "adapters") {
		m, _ := raw.(map[string]any)
		kind := getStr(m, "kind")
		known := false
		for _, k := range AdapterKinds {
			if kind == k {
				known = true
			}
		}
		if !known {
			// Hard error, never a silent skip (DR-0009).
			errs = append(errs, fmt.Sprintf("adapters[%d]: unknown kind %q", i, kind))
			continue
		}
		adp := Adapter{Kind: kind, Match: getStr(m, "match")}
		if adp.Match == "" {
			errs = append(errs, fmt.Sprintf("adapters[%d]: match is required", i))
		}
		if kind == "prefix-namespace" {
			adp.Prefix = getStr(m, "prefix")
			if adp.Prefix == "" {
				errs = append(errs, fmt.Sprintf("adapters[%d]: prefix is required", i))
			}
		} else {
			adp.FMField = getStr(m, "field")
			adp.Catalogue = map[string][]string{}
			for pname, globs := range getMap(m, "catalogue") {
				gl, _ := strList(globs)
				adp.Catalogue[pname] = gl
			}
			if adp.FMField == "" {
				errs = append(errs, fmt.Sprintf("adapters[%d]: field is required", i))
			}
		}
		adapters = append(adapters, adp)
	}

	profiles := map[string]Profile{}
	for pname, praw := range getMap(data, "profiles") {
		pm, _ := praw.(map[string]any)
		es := getMap(pm, "export_slice")
		inc, _ := strList(es["include"])
		if len(inc) == 0 {
			inc = []string{"*"}
		}
		exc, _ := strList(es["exclude"])
		profiles[pname] = Profile{Name: pname, ExportInclude: inc, ExportExclude: exc}
	}

	retracted, _ := strList(data["retracted"])
	manifestName := getStr(data, "manifest_name")
	if manifestName == "" {
		manifestName = name + "-manifest.json"
	} else if strings.ContainsAny(manifestName, `/\`) {
		// A path here would be joined into the consumer's `.vendkit/` too, so it
		// is rejected: relocate the publisher-side copy with publisher.manifest_dir.
		errs = append(errs, fmt.Sprintf("manifest_name %q must be a bare filename "+
			"(use publisher.manifest_dir to relocate the publisher-side copy)", manifestName))
	}

	known := map[string]bool{"schema_version": true, "slice": true,
		"publisher": true, "include": true, "exclude": true, "seed": true,
		"adapters": true, "profiles": true, "retracted": true,
		"manifest_name": true}
	var unknown []string
	for key := range data {
		if !known[key] {
			unknown = append(unknown, key)
		}
	}
	sort.Strings(unknown)
	for _, key := range unknown {
		errs = append(errs, fmt.Sprintf("unknown top-level key: %q", key))
	}

	if len(errs) > 0 {
		return nil, Usagef("%s: %s", path, strings.Join(errs, "; "))
	}
	return &ExportDecl{
		SliceName: name, SliceTitle: title,
		PublisherSCM: scm, PublisherRepo: repo, ManifestDir: manifestDir,
		Include: include, Exclude: exclude, Seed: seed,
		Adapters: adapters, Profiles: profiles, Retracted: retracted,
		ManifestName: manifestName, Path: path,
	}, nil
}

// -- export surface -------------------------------------------------------------

// matched: sorted repo-relative posix paths — matched(patterns) − exclude.
// Regular files only; symlinks are rejected (export-declaration spec §2).
func (d *ExportDecl) matched(root string, patterns []string) ([]string, error) {
	found := map[string]bool{}
	for _, pattern := range patterns {
		hits, err := TreeGlob(root, pattern)
		if err != nil {
			return nil, err
		}
		for _, h := range hits {
			if h.IsSymlink {
				return nil, Usagef("symlink in export surface: %s", h.Rel)
			}
			found[h.Rel] = true
		}
	}
	var out []string
	for rel := range found {
		if !MatchAny(rel, d.Exclude) {
			out = append(out, rel)
		}
	}
	sort.Strings(out)
	return out, nil
}

// ExportedFiles: the vendored (drift-gated) surface.
func (d *ExportDecl) ExportedFiles(root string) ([]string, error) {
	result, err := d.matched(root, d.Include)
	if err != nil {
		return nil, err
	}
	if len(result) == 0 && len(d.Seed) == 0 {
		return nil, Usagef("%s: export surface is empty", d.Path)
	}
	seeded, err := d.matched(root, d.Seed)
	if err != nil {
		return nil, err
	}
	inResult := map[string]bool{}
	for _, r := range result {
		inResult[r] = true
	}
	var overlap []string
	for _, s := range seeded {
		if inResult[s] {
			overlap = append(overlap, s)
		}
	}
	if len(overlap) > 0 {
		// A path cannot be both drift-gated and free-to-diverge (DR-0013).
		sort.Strings(overlap)
		if len(overlap) > 3 {
			overlap = overlap[:3]
		}
		return nil, Usagef("%s: path(s) matched by both include and seed: %s",
			d.Path, strings.Join(overlap, ", "))
	}
	return result, nil
}

// SeededFiles: the scaffold-once surface (DR-0013).
func (d *ExportDecl) SeededFiles(root string) ([]string, error) {
	return d.matched(root, d.Seed)
}

// -- adapters ---------------------------------------------------------------------

func (d *ExportDecl) AdaptersFor(path string) ([]Adapter, error) {
	var hits []Adapter
	for _, a := range d.Adapters {
		if PathMatch(path, a.Match) {
			hits = append(hits, a)
		}
	}
	for _, kind := range AdapterKinds {
		n := 0
		for _, a := range hits {
			if a.Kind == kind {
				n++
			}
		}
		if n > 1 {
			return nil, Usagef("%s: more than one %s adapter matches", path, kind)
		}
	}
	return hits, nil
}

func (d *ExportDecl) ConsumerPath(path string) (string, error) {
	hits, err := d.AdaptersFor(path)
	if err != nil {
		return "", err
	}
	for _, a := range hits {
		if a.Kind == "prefix-namespace" {
			head, tail := "", path
			if idx := strings.LastIndex(path, "/"); idx >= 0 {
				head, tail = path[:idx], path[idx+1:]
			}
			if !strings.HasPrefix(tail, a.Prefix) {
				tail = a.Prefix + tail
			}
			if head != "" {
				return head + "/" + tail, nil
			}
			return tail, nil
		}
	}
	return path, nil
}

// -- profiles ----------------------------------------------------------------------

// ProfileInScope: whether reconcile-scope may offer `path` to this profile.
func (d *ExportDecl) ProfileInScope(profile, path string) bool {
	if profile == "" {
		return true // unbound consumer takes the whole surface
	}
	prof, ok := d.Profiles[profile]
	if !ok {
		return true
	}
	return MatchAny(path, prof.ExportInclude) && !MatchAny(path, prof.ExportExclude)
}
